src/TodosContext/TodosContext.tsx

```diff
@@ -0,0 +1,390 @@
+                }
+
+                return todo;
+              }),
+            );
+            setProcessingIds(prev =>
+              prev.filter(currentTodoId => currentTodoId !== todoId),
+            );
+          })
+          .catch(() => {
+            setErrorMessage(ErrorMessage.UnableToUpdateATodo);
+            setTimeout(() => setErrorMessage(null), 3000);
+            setEditedTodoId(todoId);
+            setProcessingIds(prev =>
+              prev.filter(currentTodoId => currentTodoId !== todoId),
+            );
+          });
+      } else {
+        setEditedTodoId(null);
+      }
+    } else if (!editedTodoTitle) {
+      handleDeleteEmptyTodo(editedTodoId as number);
+    }
+  };
+
+  const handleEditKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setEditedTodoId(null);
+    }
+  };
+
+  const stopEditTodo = () => {
+    setEditedTodoId(null);
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setIsSubmiting(true);
+    setEditedTodoId(0);
+    setProcessingIds(prev => [...prev, todoId]);
+    todoService
+      .deleteTodo(todoId)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== todoId));
+        setIsSubmiting(false);
+        setProcessingIds(prev =>
+          prev.filter(currTodoId => currTodoId !== todoId),
+        );
+      })
+      .catch(() => {
+        setIsSubmiting(false);
+        setErrorMessage(ErrorMessage.UnableToDeleteATodo);
+        setTimeout(() => setErrorMessage(null), 3000);
+      });
+  };
+
+  const handleCompleteChange = (todoId: number, checked: boolean) => {
+    setProcessingIds(curr => [...curr, todoId]);
+    todoService
+      .editTodo(todoId, { completed: checked })
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.map(todo => {
+            if (todo.id === todoId) {
+              return { ...todo, completed: checked };
+            }
+
+            return todo;
+          }),
+        );
+        setEditedTodoId(null);
+        setProcessingIds(prev => prev.filter(todo => todo !== todoId));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.UnableToUpdateATodo);
+        setTimeout(() => setErrorMessage(null), 3000);
+        setProcessingIds(prev => prev.filter(todo => todo !== todoId));
+      });
+  };
+
+  const handleSetAllAsComplited = () => {
+    if (!isAllCompleted) {
+      todos.forEach(todo => {
+        if (!todo.completed) {
+          handleCompleteChange(todo.id, true);
+        }
+      });
+    } else {
+      todos.forEach(todo => {
+        handleCompleteChange(todo.id, false);
+      });
+    }
+  };
+
+  const handleClearCompleted = () => {
+    const todosToDelete = todos.filter(todo => todo.completed);
+
+    const idsOfTodosToDelete = todosToDelete.map(todo => todo.id);
+
+    setProcessingIds(prev => [...prev, ...idsOfTodosToDelete]);
+
+    todosToDelete.map(todo => {
```

Comment:
If you don't use the returned array you don't need use `map`, you can use `foreEach`